SmartDeviceLink/private/SDLMenuReplaceOperation.m

```diff
@@ -0,0 +1,465 @@
+            if (error != nil) {
+                errors[request] = error;
+            } else {
+                // Find the id of the successful request and add it from the current menu list wherever it needs to be
+                UInt32 commandId = [SDLMenuReplaceUtilities commandIdForRPCRequest:request];
+                UInt16 position = [SDLMenuReplaceUtilities positionForRPCRequest:request];
+                [SDLMenuReplaceUtilities addCellWithCellId:commandId position:position fromNewMenuList:addMenuCells toMainMenuList:weakSelf.mutableCurrentMenu];
+            }
+        } completionHandler:^(BOOL success) {
+            if (!success) {
+                SDLLogE(@"Failed to send sub menu commands: %@", errors);
+                return completionHandler([NSError sdl_menuManager_failedToUpdateWithDictionary:errors]);
+            }
+
+            SDLLogD(@"Finished sending new cells");
+            completionHandler(nil);
+        }];
+    }];
+}
+
+#pragma mark - Dynamic Menu Helpers
+
+- (NSArray<SDLMenuCell *> *)sdl_filterDeleteMenuItemsWithOldMenuItems:(NSArray<SDLMenuCell *> *)oldMenuCells basedOnStatusList:(NSArray<NSNumber *> *)oldStatusList {
+    NSMutableArray<SDLMenuCell *> *deleteCells = [[NSMutableArray alloc] init];
+    // The index of the status should correlate 1-1 with the number of items in the menu
+    // [2,0,2,0] => [A,B,C,D] = [B,D]
+    for (NSUInteger index = 0; index < oldStatusList.count; index++) {
+        if (oldStatusList[index].integerValue == SDLMenuCellUpdateStateDelete) {
+            [deleteCells addObject:oldMenuCells[index]];
+        }
+    }
+    return [deleteCells copy];
+}
+
+- (NSArray<SDLMenuCell *> *)sdl_filterAddMenuItemsWithNewMenuItems:(NSArray<SDLMenuCell *> *)newMenuCells basedOnStatusList:(NSArray<NSNumber *> *)newStatusList {
+    NSMutableArray<SDLMenuCell *> *addCells = [[NSMutableArray alloc] init];
+    // The index of the status should correlate 1-1 with the number of items in the menu
+    // [2,1,2,1] => [A,B,C,D] = [B,D]
+    for (NSUInteger index = 0; index < newStatusList.count; index++) {
+        if (newStatusList[index].integerValue == SDLMenuCellUpdateStateAdd) {
+            [addCells addObject:newMenuCells[index]];
+        }
+    }
+    return [addCells copy];
+}
+
+- (NSArray<SDLMenuCell *> *)sdl_filterKeepMenuItemsWithOldMenuItems:(NSArray<SDLMenuCell *> *)oldMenuCells basedOnStatusList:(NSArray<NSNumber *> *)keepStatusList {
+    NSMutableArray<SDLMenuCell *> *keepMenuCells = [[NSMutableArray alloc] init];
+
+    for (NSUInteger index = 0; index < keepStatusList.count; index++) {
+        if (keepStatusList[index].unsignedIntegerValue == SDLMenuCellUpdateStateKeep) {
+            [keepMenuCells addObject:oldMenuCells[index]];
+        }
+    }
+    return [keepMenuCells copy];
+}
```

Comment:
```suggestion

```

Reply:
Why do you want me to delete this? It's used in several places.